persist: remove explicit API-level tracing events

danhhz opened the pull request:
It's never been clear if these were useful and with the tracing span
coverage, they're now redundant: we have tracing spans covering all the
same places now and tracing-subscriber has the ability to emit span
ENTER events as formatted logs. This also could be used to emit EXIT
events (which would correspond to the method finishing) and so is
strictly more general than this manual event start logging.

https://docs.rs/tracing-subscriber/0.3.14/tracing_subscriber/fmt/struct.SubscriberBuilder.html#method.with_span_events

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

This with_span_events stuff isn't hooked up to anything (yet?) so it's definitely no longer as easy to opt into this as setting an env var. But it seems worthwhile (to me) to remove the duplication. Thoughts? (Marking as draft until this is resolved)

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

aljoscha commented:
I don't think I follow yet: a lot of places where the tracing events are removed don't have a span/`#[instrument]` on the method itself, so they wouldn't produce spans that can be surfaced via `with_span_events`. I'm probably looking at it wrong, though... 😅

danhhz commented:
Ah good point, I didn't actually audit all of them! The uncovered ones look mostly like stuff when persist is first getting started up. Some we could/should add #instrument coverage for (PersistClientCache::open), some are only used in testing (PersistLocation::open_locations), and others it's probably fine to lose the log line entirely (PersistClient::new, WriteHandle::builder).

I mostly wanted to get a sense for 👍 👎 from you and paul on the idea

pH14 reviewed (approved):
👍 on removing these events. I think we can rely exclusively on `#[instrument]` and explicit spans to get what we need out of tracing

danhhz commented:
TFTRs! After letting this sit, I'm still bullish on the idea. I've yet to use these to debug anything, instead usually needing to add more targeted logging locally.

Rebased (rewrote, actually, it was much faster) and added #instrument coverage for PersistClientCache::open.
